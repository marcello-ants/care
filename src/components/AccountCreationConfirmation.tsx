import { ReactElement } from 'react';
import { Icon64BrandActionSuccess } from '@care/react-icons';
import { Grid, Typography } from '@material-ui/core';
import { StatelessSelector, Pill } from '@care/react-component-lib';
import { makeStyles } from '@material-ui/core/styles';
import Header from '@/components/Header';
import { SelectorType } from '@care/react-component-lib/dist/components/Selector/StatelessSelector';

const useStyles = makeStyles((theme) => ({
  selector: {
    '& .MuiListItem-root': {
      marginBottom: 0,
    },
  },
  icon: {
    position: 'relative',
    left: '-9px',
  },
  bodyText: {
    fontWeight: 'normal',
    lineHeight: '24px',
    margin: theme.spacing(2, 0, 4),
  },
  typeContainer: {
    margin: theme.spacing(1, 0, 1),
  },

  buttonContainer: {
    padding: theme.spacing(3.5, 3, 0),
  },
  header: {
    marginBottom: ({ subheader }: { subheader: string | undefined }) =>
      subheader ? 0 : theme.spacing(1),
  },
}));

type Options = {
  label: string;
  value: string;
};

interface IAccountCreationConfirmation {
  header: string;
  description?: string;
  subheader?: string;
  options: Options[];
  onChange?: SelectorType['onChange'];
  cta?: ReactElement;
  value?: string;
  showIcon?: boolean;
}

function AccountCreationConfirmation({
  header,
  description,
  subheader,
  options,
  value,
  onChange,
  cta,
  showIcon,
}: IAccountCreationConfirmation) {
  const classes = useStyles({ subheader });

  return (
    <Grid container>
      {showIcon && (
        <Grid item xs={12}>
          {/*
          The svg has a litle padding that makes the icon smaller
          60px makes the icon size to 49.33 the correct size is 60px approx
          74px makes the icon size 60.88
        */}
          <Icon64BrandActionSuccess size="74px" className={classes.icon} />
        </Grid>
      )}

      <Grid item xs={12} className={classes.header}>
        <Header>{header}</Header>
      </Grid>
      {description && (
        <Grid item xs={12}>
          <Typography variant="body2" className={classes.bodyText}>
            {description}
          </Typography>
        </Grid>
      )}

      {subheader && (
        <Grid item xs={12}>
          <Typography variant="h4">{subheader}</Typography>
        </Grid>
      )}

      <Grid item xs={12} className={classes.typeContainer}>
        <StatelessSelector onChange={onChange} single value={value} className={classes.selector}>
          {options.map((option) => (
            <Pill key={option.label} {...option} size="md" />
          ))}
        </StatelessSelector>
        {cta && <div className={classes.buttonContainer}>{cta}</div>}
      </Grid>
    </Grid>
  );
}

AccountCreationConfirmation.defaultProps = {
  value: '',
  onChange: () => {},
  cta: null,
  showIcon: true,
  description: '',
  subheader: '',
};

export default AccountCreationConfirmation;
