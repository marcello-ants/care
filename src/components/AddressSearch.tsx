import { useEffect, useState } from 'react';
import { theme } from '@care/material-ui-theme';
import { Box, Button, makeStyles } from '@material-ui/core';
import { GoogleMaps, GoogleMapsActions, useGoogleMaps } from '@care/seeker-react-components';
import { Icon24UtilitySearch } from '@care/react-icons';
import { InlineTextField, Typography } from '@care/react-component-lib';
import { AddressInput } from '@/__generated__/globalTypes';
import { logIbAddressErrorEvent } from '@/utilities/analyticsPagesUtils';

const useStyles = makeStyles(() => ({
  addressSearchContainer: {
    position: 'relative',
    marginTop: theme.spacing(-2.75),
    '&& > .MuiTextField-root': {
      paddingLeft: 0,
      paddingRight: 0,
    },
    '&& label': {
      paddingLeft: theme.spacing(5),
      textAlign: 'left',
    },
    '&& input': {
      paddingLeft: theme.spacing(7),
    },
  },
  loopIcon: {
    position: 'absolute',
    top: 42,
    left: 20,
    zIndex: 2,
  },
  locationContainer: {
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'space-between',
  },
  locationText: {
    textAlign: 'left',
    whiteSpace: 'pre-wrap',
  },
  editLocationButton: {
    width: 81,
  },
  notFoundContainer: {
    padding: theme.spacing(5, 0),
    textAlign: 'center',
  },
  suiteContainer: {
    '&& > .MuiTextField-root': {
      paddingLeft: 0,
      paddingRight: 0,
    },
  },
}));

interface AddressSearchProps {
  seekerAddress: AddressInput;
  setSeekerAddress: (address: AddressInput) => void;
  displayAddress: string;
  setDisplayAddress: (displayAddress: string) => void;
  pageName: string;
}

const AddressSearch = ({
  seekerAddress,
  setSeekerAddress,
  displayAddress,
  setDisplayAddress,
  pageName,
}: AddressSearchProps): JSX.Element => {
  const classes = useStyles();
  const [{ placeDetails, predictions }, dispatch] = useGoogleMaps();
  const addressLabelText = 'Street address';
  const addressLabelTextOnInput = 'Search address';
  const suiteLabelText = 'Apt or Suite (optional)';
  const invalidAddressText = 'Invalid address';
  const notFoundText = 'Address not found';
  const notFoundDescription = `Make sure it's spelled correctly or try adding a city, state or zip code`;
  const CITY_POSITION: number = 2;
  const STATE_POSITION: number = 3;

  const [label, setLabel] = useState(addressLabelText);
  const [showAddressSearch, setShowAddressSearch] = useState<boolean>(true);
  const [shouldSaveAddress, setShouldSaveAddress] = useState<boolean>(false);

  const isValidAddress = seekerAddress.zip && seekerAddress.addressLine1;

  const handleFocus = () => {
    setLabel(addressLabelTextOnInput);
  };

  const handleBlur = () => {
    setLabel(addressLabelText);
  };

  const setGoogleMapsPlaceInfo = (inputText: string) => {
    dispatch({ type: GoogleMapsActions.SET_INPUT_SEARCH, payload: inputText });
    dispatch({ type: GoogleMapsActions.SET_PLACE_DETAILS, payload: null });
    dispatch({ type: GoogleMapsActions.CLEAR_PREDICTIONS });
  };

  const handleEditClick = () => {
    setGoogleMapsPlaceInfo(seekerAddress.addressLine1);
    setShowAddressSearch(true);
    setShouldSaveAddress(false);
  };

  const handlePredictionClick = () => {
    setShowAddressSearch(false);
    setShouldSaveAddress(true);
  };

  useEffect(() => {
    if (shouldSaveAddress && placeDetails) {
      setSeekerAddress({
        ...seekerAddress,
        addressLine1: placeDetails.prediction.structured_formatting.main_text,
        city: placeDetails.prediction.terms[CITY_POSITION].value,
        state: placeDetails.prediction.terms[STATE_POSITION].value,
        zip: placeDetails.zip || '',
        longitude: placeDetails.longitude,
        latitude: placeDetails.latitude,
      });
      setDisplayAddress(
        `${placeDetails.prediction.structured_formatting.main_text} \n${
          placeDetails.prediction.structured_formatting.secondary_text
        } ${placeDetails.zip || ''}`
      );
      setShouldSaveAddress(false);
    }
  }, [shouldSaveAddress, placeDetails]);

  useEffect(() => {
    if (displayAddress) {
      setShowAddressSearch(false);
    }
  }, []);

  const handleSuiteChange = (value: string) => {
    setSeekerAddress({
      ...seekerAddress,
      addressLine2: value,
    });
  };

  const addressSearchMode = showAddressSearch || shouldSaveAddress;

  useEffect(() => {
    if (addressSearchMode && !predictions) {
      logIbAddressErrorEvent(notFoundText, pageName);
    }
  }, [addressSearchMode, predictions]);
  useEffect(() => {
    if (!addressSearchMode && !isValidAddress) {
      logIbAddressErrorEvent(invalidAddressText, pageName);
    }
  }, [addressSearchMode, isValidAddress]);

  return (
    <Box>
      {addressSearchMode ? (
        <Box className={classes.addressSearchContainer}>
          <Icon24UtilitySearch className={classes.loopIcon} />
          <GoogleMaps.InputSearch label={label} onFocus={handleFocus} onBlur={handleBlur} />
          {predictions ? (
            predictions.map((prediction) => (
              <GoogleMaps.Prediction
                {...prediction}
                key={prediction.place_id}
                onClick={handlePredictionClick}
              />
            ))
          ) : (
            <Box className={classes.notFoundContainer}>
              <Typography variant="h2">{notFoundText}</Typography>
              <Typography variant="body1">{notFoundDescription}</Typography>
            </Box>
          )}
        </Box>
      ) : (
        <Box>
          <Box className={classes.locationContainer}>
            <Box className={classes.locationText}>
              <Typography>{displayAddress}</Typography>
            </Box>
            <Box>
              <Button
                color="secondary"
                variant="outlined"
                size="small"
                className={classes.editLocationButton}
                onClick={handleEditClick}>
                Edit
              </Button>
            </Box>
          </Box>
          {!isValidAddress && (
            <Box mt={2}>
              <Typography variant="body2" color="error">
                {invalidAddressText}
              </Typography>
            </Box>
          )}
          <Box className={classes.suiteContainer}>
            <InlineTextField
              id="aptOrSuite"
              InputLabelProps={{ shrink: false }}
              name="aptOrSuite"
              label={suiteLabelText}
              value={seekerAddress.addressLine2}
              onChange={(e) => handleSuiteChange(e.target.value)}
            />
          </Box>
        </Box>
      )}
    </Box>
  );
};

const AddressSearchWrapper = (props: AddressSearchProps) => {
  return (
    <GoogleMaps.Provider>
      <AddressSearch {...props} />
    </GoogleMaps.Provider>
  );
};

export default AddressSearchWrapper;
