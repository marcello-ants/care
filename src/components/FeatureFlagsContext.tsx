import { createContext, useContext, ReactNode, useState } from 'react';
import { LDEvaluationDetail } from 'launchdarkly-node-server-sdk';

interface FeatureFlagsProviderProps {
  children: ReactNode;
  flags: FeatureFlags;
}

export type featureFlagsState = {
  flags: FeatureFlags;
};

export type FeatureFlags = Record<string, LDEvaluationDetail | undefined>;

const initialState: featureFlagsState = {
  flags: {},
};

const FeatureFlagsContext = createContext<featureFlagsState>(initialState);

export const FeatureFlagsProvider = ({ children, flags }: FeatureFlagsProviderProps) => {
  const [state] = useState({ flags });
  return <FeatureFlagsContext.Provider value={state}>{children}</FeatureFlagsContext.Provider>;
};

export const useFeatureFlags = () => useContext(FeatureFlagsContext);
