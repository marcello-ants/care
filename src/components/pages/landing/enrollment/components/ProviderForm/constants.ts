export const PROVIDER_SERVICES_DESKTOP = [
  {
    description: 'Child care',
    label: 'Child care',
    value: 'CHILD_CARE',
  },
  {
    description: 'Senior care',
    label: 'Senior care',
    value: 'SENIOR_CARE',
  },
  {
    description: 'Housekeeping',
    label: 'Housekeeping',
    value: 'HOUSEKEEPING',
  },
  {
    description: 'Pet care',
    label: 'Pet care',
    value: 'PET_CARE',
  },
  {
    description: 'Tutoring',
    label: 'Tutoring',
    value: 'TUTORING',
  },
  {
    description: 'Special needs care',
    label: 'Special needs care',
    value: 'SPECIAL_NEEDS',
  },
  {
    description: 'Errands & odd jobs',
    label: 'Errands & odd jobs',
    value: 'GENERIC',
  },
];

export const PROVIDER_SERVICES_MOBILE = [
  {
    description: 'Child care',
    label: 'Child care',
    value: 'CHILD_CARE',
  },
  {
    description: 'Senior care',
    label: 'Senior care',
    value: 'SENIOR_CARE',
  },
  {
    description: 'Pet care',
    label: 'Pet care',
    value: 'PET_CARE',
  },
  {
    description: 'Tutoring',
    label: 'Tutoring',
    value: 'TUTORING',
  },
  {
    description: 'Special needs care',
    label: 'Special needs care',
    value: 'SPECIAL_NEEDS',
  },
  {
    description: 'Housekeeping',
    label: 'Housekeeping',
    value: 'HOUSEKEEPING',
  },
  {
    description: 'Errands & odd jobs',
    label: 'Errands & odd jobs',
    value: 'GENERIC',
  },
];

export const DROPOFF_URLS = {
  CHILD_CARE: '/swf/ccOptimizedFlow',
  SENIOR_CARE: '/app/enrollment/provider/sc ',
  PET_CARE: '/swf/pcOptimizedFlow',
  HOUSEKEEPING: '/swf/hkOptimizedFlow',
  TUTORING: '/swf/tuOptimizedFlow',
  SPECIAL_NEEDS: '/swf/snOptimizedFlow',
  GENERIC: '/swf/cgOptimizedFlow',
  AU_PAIR: '',
  CARE_GIGS: '',
};
