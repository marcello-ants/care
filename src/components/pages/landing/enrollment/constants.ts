export const SITTER_FORM = {
  providerServices: {
    title: 'What Services do you offer?',
    id: 'providerService',
    options: [
      {
        value: 'CHILD_CARE',
        text: 'Babysitter',
      },
      {
        value: 'CHILD_CARE',
        text: 'Nanny',
      },
      {
        value: 'SPECIAL_NEEDS',
        text: 'Special Needs Provider',
      },
      {
        value: 'TUTORING',
        text: 'Tutor',
      },
      {
        value: 'TUTORING',
        text: 'Private Lesson Instructor',
      },
      {
        value: 'SENIOR_CARE',
        text: 'Senior Care Provider',
      },
      {
        value: 'PET_CARE',
        text: 'Pet Care Provider',
      },
      {
        value: 'HOUSEKEEPING',
        text: 'Housekeeper',
      },
      {
        value: 'GENERIC',
        text: 'Errands &amp; Odd Jobs Provider',
      },
    ],
  },
  providerEmail: {
    id: 'email',
    placeholder: 'Email',
  },
  providerPassword: {
    id: 'password',
    placeholder: 'Password',
  },
  providerZip: {
    id: 'providerZip',
    placeholder: 'ZIP',
  },
};

export const GENERIC_FORM = {
  howDidYouHearAboutUs: {
    title: 'How did you hear about us?',
    id: 'howDidYouHearAboutUs',
    options: [
      {
        value: 'FACEBOOK',
        text: 'Facebook',
      },
      {
        value: 'BANNER_AD',
        text: 'Banner Ad',
      },
      {
        value: 'TV_AD',
        text: 'TV Ad',
      },
      {
        value: 'OTHER_SOCIAL_MEDIA',
        text: 'Other Social Media (Twitter, Instagram, Pinterest, LinkedIn)',
      },
      {
        value: 'PRESS_COVERAGE',
        text: 'Press Coverage (News, Magazine, Blog)',
      },
      {
        value: 'FRIENDS_OR_FAMILY',
        text: 'Friends or Family',
      },
      {
        value: 'ONLINE_VIDEO',
        text: 'Online Video (Hulu, Youtube)',
      },
      {
        value: 'SEARCH_ENGINE',
        text: 'Search Engine (Google, Bing)',
      },
      {
        value: 'PARENTING_GROUP',
        text: 'Parenting Group or Forum',
      },
      {
        value: 'OTHER',
        text: 'Other',
      },
    ],
  },
};
