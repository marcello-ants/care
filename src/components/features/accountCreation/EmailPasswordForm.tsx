import React, { useEffect } from 'react';
import { Button, Grid } from '@material-ui/core';
import { makeStyles } from '@material-ui/core/styles';
import getConfig from 'next/config';
import { useFormikContext } from 'formik';
import { HearAboutUsField } from '@/components/HearAboutUs';
import { useFeatureFlags } from '@/components/FeatureFlagsContext';
import useEnterKey from '@/components/hooks/useEnterKey';
import { CLIENT_FEATURE_FLAGS } from '@/constants';
import { AnalyticsHelper } from '@/utilities/analyticsHelper';
import LegalDisclaimer from '@/components/features/accountCreation/LegalDisclaimer';
import EmailField from './EmailField';
import PasswordField from './PasswordField';
import { FormValues } from './accountCreationForm';

const useStyles = makeStyles((theme) => ({
  terms: {
    width: '200px',
    textAlign: 'center',
    margin: theme.spacing(1, 'auto', 0),
  },
  hearAboutUsContainer: {
    '& > div': {
      paddingRight: 0,
      paddingLeft: 0,
      paddingBottom: 0,
    },
  },
  joinNowButtonContainer: {
    padding: theme.spacing(4, 3, 3),
  },
  inputContainer: {
    '& > div': {
      padding: 0,
    },
  },
}));

const {
  publicRuntimeConfig: { CZEN_GENERAL_LOGIN },
} = getConfig();

type EmailPasswordFormProps = {
  handleJoinNow: any;
};

function EmailPasswordForm(props: EmailPasswordFormProps) {
  const classes = useStyles();
  const { handleJoinNow } = props;

  const featureFlags = useFeatureFlags();
  const hdyhauFlagEvaluation = featureFlags?.flags[CLIENT_FEATURE_FLAGS.HDYHAU];
  const showHDYHAU = hdyhauFlagEvaluation?.variationIndex === 1;

  const { isValidating, errors } = useFormikContext<FormValues>();
  const buttonDisabled =
    isValidating || Boolean(errors.email || errors.password || errors.howDidYouHearAboutUs);

  useEnterKey(!buttonDisabled, handleJoinNow);

  const passwordCopyTestFlag = featureFlags?.flags[CLIENT_FEATURE_FLAGS.PASSWORD_REQUIREMENTS_COPY];
  const passwordCopyTestValue = passwordCopyTestFlag?.value;

  const isPasswordCopyTest = passwordCopyTestValue && [2, 3].includes(passwordCopyTestValue);

  useEffect(() => {
    AnalyticsHelper.logTestExposure(
      CLIENT_FEATURE_FLAGS.PASSWORD_REQUIREMENTS_COPY,
      passwordCopyTestFlag
    );
  }, []);

  useEffect(() => {
    AnalyticsHelper.logTestExposure(CLIENT_FEATURE_FLAGS.HDYHAU, hdyhauFlagEvaluation);
  }, []);

  return (
    <Grid container>
      <Grid item xs={12}>
        <div className={classes.inputContainer}>
          <EmailField
            id="email"
            name="email"
            label="Email address"
            czenGeneralLoginUrl={CZEN_GENERAL_LOGIN}
          />
        </div>
      </Grid>
      <Grid item xs={12} className={classes.inputContainer}>
        <PasswordField
          id="password"
          name="password"
          label="Password"
          minLength={isPasswordCopyTest ? 8 : undefined}
          passwordCopyTest={passwordCopyTestValue}
        />
      </Grid>
      {showHDYHAU && (
        <Grid item xs={12} className={classes.hearAboutUsContainer}>
          <HearAboutUsField name="howDidYouHearAboutUs" />
        </Grid>
      )}

      <Grid item xs={12}>
        <LegalDisclaimer cta="Continue" />
      </Grid>
      <Grid item xs={12}>
        <div className={classes.joinNowButtonContainer}>
          <Button
            color="primary"
            variant="contained"
            size="large"
            fullWidth
            disabled={buttonDisabled}
            onClick={handleJoinNow}>
            Continue
          </Button>
        </div>
      </Grid>
    </Grid>
  );
}

export default EmailPasswordForm;
