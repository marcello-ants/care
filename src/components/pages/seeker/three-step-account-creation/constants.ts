import * as yup from 'yup';
import { ServiceType } from '@/__generated__/globalTypes';
import { VerticalsAbbreviation } from '@/constants';

export const validationSchema = yup.object({
  firstName: yup
    .string()
    .min(2, 'First name needs to have at least two letters.')
    .required('First name is required'),
  lastName: yup
    .string()
    .min(2, 'Last name needs to have at least two letters.')
    .required('Last name is required'),
});

export const SERVICE_TYPE_BY_VERTICAL: { [key: string]: ServiceType } = {
  HK: ServiceType.HOUSEKEEPING,
  PC: ServiceType.PET_CARE,
  TU: ServiceType.TUTORING,
  CC: ServiceType.CHILD_CARE,
  DC: ServiceType.CHILD_CARE,
  IB: ServiceType.CHILD_CARE,
  SC: ServiceType.SENIOR_CARE,
};

export const ERROR_CREATING_ACCOUNT = 'An error occurred creating your account, please try again.';

export const FACEBOOK_VERTICALS_ALLOWANCE = ['CC'];
export const HDYHAU_VERTICALS_ALLOWANCE: VerticalsAbbreviation[] = ['HK', 'TU', 'PC', 'DC'];
