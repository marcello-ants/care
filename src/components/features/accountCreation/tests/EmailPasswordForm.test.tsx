import React from 'react';
import { render, fireEvent, screen, waitFor } from '@testing-library/react';
import { MockedProvider, MockedResponse } from '@apollo/client/testing';
import { Formik } from 'formik';

import { FeatureFlags, FeatureFlagsProvider } from '@/components/FeatureFlagsContext';
import { VALIDATE_MEMBER_EMAIL, VALIDATE_MEMBER_PASSWORD } from '@/components/request/GQL';
import { CLIENT_FEATURE_FLAGS } from '@/constants';
import EmailPasswordForm from '../EmailPasswordForm';

jest.mock('next/config', () => () => {
  return {
    publicRuntimeConfig: {
      CZEN_GENERAL_LOGIN: 'https://www.care.com/login',
    },
  };
});

const passingEmail = 'test@test.com';
const passingPassword = 'ABCDEFGH';

const emailValidationMock = {
  request: {
    query: VALIDATE_MEMBER_EMAIL,
    variables: {
      email: passingEmail,
    },
  },
  result: {
    data: {
      validateMemberEmail: { errors: [] },
    },
  },
};

const passwordValidationMock = (matchPassword: string) => {
  return {
    request: {
      query: VALIDATE_MEMBER_PASSWORD,
      variables: {
        password: matchPassword,
      },
    },
    result: {
      data: {
        validateMemberPassword: { errors: [] },
      },
    },
  };
};

let joinNowBtn: HTMLElement;
let emailField: HTMLElement;
let passwordField: HTMLElement;
let hdyhauField: HTMLElement;
let handleJoinNow: jest.Mock;

interface RenderComponentOptions {
  mocks?: ReadonlyArray<MockedResponse>;
  ldFlags?: FeatureFlags;
}

function renderComponent({ mocks, ldFlags }: RenderComponentOptions = {}) {
  handleJoinNow = jest.fn();
  const view = render(
    <MockedProvider mocks={mocks} addTypename={false}>
      <FeatureFlagsProvider flags={ldFlags || {}}>
        <Formik
          initialValues={{
            email: '',
            password: '',
          }}
          onSubmit={() => {}}
          validateOnMount>
          <EmailPasswordForm handleJoinNow={handleJoinNow} />
        </Formik>
      </FeatureFlagsProvider>
    </MockedProvider>
  );
  joinNowBtn = screen.getByRole('button', { name: 'Continue' });
  emailField = screen.getByLabelText('Email address');
  passwordField = screen.getByLabelText('Password');

  if (ldFlags?.[CLIENT_FEATURE_FLAGS.HDYHAU]?.variationIndex === 1) {
    hdyhauField = screen.getByRole('button', { name: 'How did you hear about us?' });
  }

  return view;
}

describe('EmailPasswordForm', () => {
  it('matches snapshot', () => {
    const view = renderComponent();
    expect(view.asFragment()).toMatchSnapshot();
  });

  it('should render a "Join now" button thats initially disabled', () => {
    renderComponent();
    expect(joinNowBtn).toBeDisabled();
  });

  it('should enable the "Join now" button when a valid email and password are entered', async () => {
    renderComponent({ mocks: [emailValidationMock, passwordValidationMock(passingPassword)] });
    fireEvent.change(emailField, { target: { value: passingEmail } });
    fireEvent.change(passwordField, { target: { value: passingPassword } });

    await waitFor(() => {
      expect(joinNowBtn).toBeEnabled();
    });
  });

  it('should disable the "Join now" button when an invalid email is entered', async () => {
    renderComponent({ mocks: [emailValidationMock, passwordValidationMock(passingPassword)] });
    fireEvent.change(emailField, { target: { value: passingEmail } });
    fireEvent.change(passwordField, { target: { value: passingPassword } });

    await waitFor(() => {
      expect(joinNowBtn).toBeEnabled();
    });

    fireEvent.change(emailField, { target: { value: 'email' } });

    await waitFor(() => {
      expect(joinNowBtn).toBeDisabled();
    });
  });

  it('should disable the "Join now" button when an invalid password is entered', async () => {
    renderComponent({ mocks: [emailValidationMock, passwordValidationMock(passingPassword)] });
    fireEvent.change(emailField, { target: { value: passingEmail } });
    fireEvent.change(passwordField, { target: { value: passingPassword } });

    await waitFor(() => {
      expect(joinNowBtn).toBeEnabled();
    });

    fireEvent.change(passwordField, { target: { value: 'foo' } });

    await waitFor(() => {
      expect(joinNowBtn).toBeDisabled();
    });
  });

  it('should call props.handleJoinNow when the join now button is clicked', async () => {
    renderComponent({ mocks: [emailValidationMock, passwordValidationMock(passingPassword)] });
    fireEvent.change(emailField, { target: { value: passingEmail } });
    fireEvent.change(passwordField, { target: { value: passingPassword } });

    await waitFor(() => {
      expect(joinNowBtn).toBeEnabled();
    });

    fireEvent.click(joinNowBtn);
    expect(handleJoinNow).toHaveBeenCalled();
  });

  describe('with HDYHAU enabled', () => {
    const ldFlags: FeatureFlags = {
      [CLIENT_FEATURE_FLAGS.HDYHAU]: {
        reason: { kind: '' },
        value: 1,
        variationIndex: 1,
      },
    };

    it('should render the HDYHAU select when the flag is enabled', () => {
      renderComponent({
        ldFlags,
        mocks: [emailValidationMock, passwordValidationMock(passingPassword)],
      });
      expect(hdyhauField).toBeVisible();
    });

    it('should enable the "Join now" button when all inputs are entered', async () => {
      renderComponent({
        ldFlags,
        mocks: [emailValidationMock, passwordValidationMock(passingPassword)],
      });
      fireEvent.change(emailField, { target: { value: passingEmail } });
      fireEvent.change(passwordField, { target: { value: passingPassword } });
      const hdyhauInput = hdyhauField.parentElement?.querySelector('input');
      fireEvent.change(hdyhauInput!, { target: { value: 'OTHER' } });

      await waitFor(() => {
        expect(joinNowBtn).toBeEnabled();
      });
    });

    it('should disable the "Join now" button if HDYHAU is not entered', async () => {
      renderComponent({
        ldFlags,
        mocks: [emailValidationMock, passwordValidationMock(passingPassword)],
      });
      fireEvent.change(emailField, { target: { value: passingEmail } });
      fireEvent.change(passwordField, { target: { value: passingPassword } });
      const hdyhauInput = hdyhauField.parentElement?.querySelector('input');
      fireEvent.change(hdyhauInput!, { target: { value: 'OTHER' } });

      await waitFor(() => {
        expect(joinNowBtn).toBeEnabled();
      });

      fireEvent.change(hdyhauInput!, { target: { value: '' } });

      await waitFor(() => {
        expect(joinNowBtn).toBeDisabled();
      });
    });
  });
});
