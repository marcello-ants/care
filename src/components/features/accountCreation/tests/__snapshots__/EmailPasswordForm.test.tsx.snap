// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmailPasswordForm matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root MuiGrid-container"
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <div
        class="makeStyles-inputContainer-4"
      >
        <div
          class="MuiFormControl-root MuiTextField-root makeStyles-inline-5"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
            data-shrink="false"
            for="email"
            id="email-label"
          >
            Email address
          </label>
          <div
            class="MuiInputBase-root MuiInput-root MuiInputBase-formControl MuiInput-formControl"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiInput-input"
              id="email"
              name="email"
              type="email"
              value=""
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root makeStyles-inputContainer-4 MuiGrid-item MuiGrid-grid-xs-12"
    >
      <div
        class="MuiFormControl-root MuiTextField-root makeStyles-inline-5"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
          data-shrink="false"
          for="password"
          id="password-label"
        >
          Password
        </label>
        <div
          class="MuiInputBase-root MuiInput-root MuiInputBase-formControl MuiInput-formControl"
        >
          <input
            aria-describedby="password-helper-text"
            aria-invalid="false"
            class="MuiInputBase-input MuiInput-input"
            id="password"
            name="password"
            type="password"
            value=""
          />
        </div>
        <p
          class="MuiFormHelperText-root"
          id="password-helper-text"
        >
          Minimum 6 characters
        </p>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <div
        class="makeStyles-terms-9 makeStyles-terms-11"
      >
        <p
          class="MuiTypography-root makeStyles-info1-16 MuiTypography-body1"
        >
          <span>
            By clicking "
          </span>
          Continue
          <span>
            ", you agree to ourÂ 
          </span>
          <a
            href="https://www.care.com/terms-of-use-popup-p1030.html"
            rel="noreferrer"
            tabindex="-1"
            target="_blank"
          >
            Terms of Use
          </a>
           
          <span>
            and
          </span>
           
          <a
            href="https://www.care.com/privacy-policy-popup-p1031.html"
            rel="noreferrer"
            tabindex="-1"
            target="_blank"
          >
            Privacy Policy
          </a>
          <span>
            .
          </span>
        </p>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <div
        class="makeStyles-joinNowButtonContainer-3"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-containedSizeLarge MuiButton-sizeLarge Mui-disabled MuiButton-fullWidth Mui-disabled"
          disabled=""
          tabindex="-1"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Continue
          </span>
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
